Userland/Services/DeviceMapper/DeviceEventLoop.cpp

```diff
@@ -0,0 +1,241 @@
+    : m_devctl_file(move(devctl_file))
+{
+}
+
+using MinorNumberAllocationType = DeviceEventLoop::MinorNumberAllocationType;
+
+static constexpr DeviceEventLoop::DeviceNodeMatch s_matchers[] = {
+    { "audio"sv, "audio"sv, "audio/%digit"sv, false, 116, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0220 },
+    { {}, "render"sv, "gpu/render%digit"sv, false, 28, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "window"sv, "gpu-connector"sv, "gpu/connector%digit"sv, false, 226, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0660 },
+    { {}, "virtio-console"sv, "hvc0p%digit"sv, false, 229, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "phys"sv, "hid-mouse"sv, "input/mouse/%digit"sv, false, 10, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { "phys"sv, "hid-keyboard"sv, "input/keyboard/%digit"sv, false, 85, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0666 },
+    { {}, "storage"sv, "hd%letter"sv, true, 3, MinorNumberAllocationType::SequentialUnlimited, 0, 0, 0600 },
+    { "tty"sv, "console"sv, "tty%digit"sv, false, 4, MinorNumberAllocationType::SequentialLimited, 0, 63, 0620 },
+    { "tty"sv, "console"sv, "ttyS%digit"sv, false, 4, MinorNumberAllocationType::SequentialLimited, 64, 127, 0620 },
+};
+
+static bool is_in_minor_number_range(DeviceEventLoop::DeviceNodeMatch const& matcher, MinorNumber minor_number)
+{
+    if (matcher.minor_number_allocation_type == MinorNumberAllocationType::SequentialUnlimited)
+        return true;
+
+    return matcher.minor_number_start <= minor_number && static_cast<MinorNumber>(matcher.minor_number_start.value() + matcher.minor_number_range_size) >= minor_number;
+}
+
+static Optional<DeviceEventLoop::DeviceNodeMatch const&> device_node_family_to_match_type(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    for (auto& matcher : s_matchers) {
+        if (matcher.major_number == major_number
+            && block_device == matcher.block_device
+            && is_in_minor_number_range(matcher, minor_number))
+            return matcher;
+    }
+    return {};
+}
+
+static bool is_in_family_minor_number_range(DeviceNodeFamily const& family, MinorNumber minor_number)
+{
+    return family.base_minor_number() <= minor_number && static_cast<MinorNumber>(family.base_minor_number().value() + family.devices_symbol_suffix_allocation_map().size()) >= minor_number;
+}
+
+Optional<DeviceNodeFamily&> DeviceEventLoop::find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const
+{
+    DeviceNodeFamily::Type type = block_device ? DeviceNodeFamily::Type::CharacterDevice : DeviceNodeFamily::Type::CharacterDevice;
+    for (auto const& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
```

Comment:
Code style: instead of `type == family->type()`, use `family->type() == type` so it's consistent with `family->major_number() == major_number`